edx_notifications/stores/sql/store_provider.py

```diff
@@ -0,0 +1,283 @@
+
+        query = SQLNotificationUserMap.objects.filter(user_id=user_id)
+
+        if select_related:
+            query = query.select_related()
+
+        if msg_id:
+            query = query.filter(msg__id=msg_id)
+
+        if namespace:
+            query = query.filter(msg__namespace=namespace)
+
+        if not (read and unread):
+            if read:
+                query = query.filter(read_at__isnull=False)
+
+            if unread:
+                query = query.filter(read_at__isnull=True)
+
+        if type_name:
+            query = query.filter(msg__msg_type=type_name)
+
+        return query[offset:offset + limit]
+
+    def get_num_notifications_for_user(self, user_id, filters=None):
+        """
+        Returns an integer count of edx_notifications. It is presumed
+        that store provider implementations can make this an optimized
+        query
+
+        ARGS:
+            - user_id: The id of the user
+            - filters: a dict containing
+                - namespace: what namespace to search (defuault None)
+                - read: Whether to return read notifications (default True)
+                - unread: Whether to return unread notifications (default True)
+                - type_name: which type to return
+
+        RETURNS: integer
+
+        NOTE: You cannot pass in a filter of 'msg_id'
+        """
+
+        if filters and 'msg_id' in filters:
+            raise ValueError(
+                'It is non-sensical to pass in a msg_id '
+                'filter parameter into an aggregate counting method!'
+            )
+
+        return self._get_notifications_for_user(
+            user_id,
+            filters=filters,
+        ).count()
+
+    def get_notifications_for_user(self, user_id, filters=None, options=None):
+        """
+        Returns a (unsorted) collection (list) of notifications for the user.
+
+        NOTE: We will have to add paging (with sorting/filtering) in the future
+
+        ARGS:
+            - user_id: The id of the user
+            - filters: a dict containing
+                - msg_id: a particular notification id to retrieve
+                    ( if this is used in conjuction with other filters, it may
+                      result in an empty set being returned )
+                - namespace: what namespace to search (defuault None)
+                - read: Whether to return read notifications (default True)
+                - unread: Whether to return unread notifications (default True)
+                - type_name: which type to return
+            - options: a dict containing some optional parameters
+                - limit: max number to return (up to some system defined max)
+                - offset: offset into the list, to implement paging
+
+        RETURNS: list   i.e. []
+        """
+
+        _options = copy.deepcopy(options) if options else {}
+        _options['select_related'] = True
```

Comment:
You don't need a `deepcopy` here. You're only modifying the top of the dict, so a regular `.copy()` would be enough.